### PR TITLE
[SPARK-19916][SQL] simplify bad file handling

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

We should only have one centre place to try catch the exception for corrupted files.

## How was this patch tested?

existing test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sameeragarwal @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 1]
```diff
@@ -130,54 +144,35 @@ class FileScanRDD(
...
+                try {
+                  if (internalIter.hasNext) {
+                    internalIter.next()
+                  } else {
+                    finished = true
+                    null
                   }
+                } catch {
+                  // Throw FileNotFoundException even `ignoreCorruptFiles` is true
+                  case e: java.io.FileNotFoundException => throw e
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: nit: `FileNotFoundException` will be thrown anyway, do wee need this case?

**Comment by viirya** [comment]
LGTM with very minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 2]
```diff
@@ -44,7 +44,7 @@ case class PartitionedFile(
     filePath: String,
     start: Long,
     length: Long,
-    locations: Array[String] = Array.empty) {
+    @transient locations: Array[String] = Array.empty) {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: do we need to mark it as `transient`? `filePartitions: Seq[FilePartition])` is already `transient` in `FileScanRDD`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: `FileNotFoundException` extends `IOException`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: this is not for `FileScanRDD.filePartitions`, this is for `FilePartition`s that sent by scheduler. The location is only useful during planning, we should not send it to executors.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 3]
```diff
@@ -90,16 +90,14 @@ trait FileFormat {
...
-  def buildReader(
+  protected def buildReader(
       sparkSession: SparkSession,
       dataSchema: StructType,
       partitionSchema: StructType,
       requiredSchema: StructType,
       filters: Seq[Filter],
       options: Map[String, String],
       hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
-    // TODO: Remove this default implementation when the other formats have been ported
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: No more TODO here?

**Comment by viirya** [comment]
LGTM

**Review by gatorsmile — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 4]
```diff
@@ -130,54 +144,35 @@ class FileScanRDD(
...
+                    internalIter.next()
+                  } else {
+                    finished = true
+                    null
                   }
+                } catch {
+                  // Throw FileNotFoundException even `ignoreCorruptFiles` is true
+                  case e: java.io.FileNotFoundException => throw e
+                  case e @ (_: RuntimeException | _: IOException) =>
+                    logWarning(s"Skipped the rest content in the corrupted file: $currentFile", e)
```
Comment (apurvis) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Better English: "Skipped the rest _of the_ content in the corrupted file:"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 5]
```diff
@@ -121,6 +121,20 @@ class FileScanRDD(
         nextElement
       }
 
+      private def readCurrentFile(): Iterator[InternalRow] = {
+        try {
+          readFunction(currentFile)
+        } catch {
+          case e: java.io.FileNotFoundException =>
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Why not import the class? same below

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: Actually we don't need to implement this method in all sub-classes. Some `FileFormat` may override `buildReaderWithPartitionValues` directly(parquet), Some `FileFormat` may not be used in read path(HiveFileFormat)

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 3 above)
Comment (gkumar7) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: @cloud-fan I am wondering why the `buildReader` method is now marked as `protected`? Maybe you can comment here: https://issues.apache.org/jira/browse/SPARK-27751

**Pull request closed without merging**